Rewrite ar and dr with newshell

thestr4ng3r opened the pull request:
**DO NOT SQUASH**

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

`ar` and `dr` now share exactly the same commands (except for `drx`, which is debug-specific):

![Bildschirmfoto 2021-12-05 um 17 58 38](https://user-images.githubusercontent.com/1460997/144755848-54123a6d-ecb7-45cd-b153-3af26d014fd2.png)

Breaking changes:
ar* => arf, ar- => arf-, ar. => ar*
ar, => art, ar? => arq
ar0 => ara0, arb => arab, ars => araS
arC => arpC
arn => arR
art => arT
arf => arF
Removed: drm, aer

**Test plan**

See the tests. 

**Closing issues**

Partially addresses https://github.com/rizinorg/rizin/issues/1442

thestr4ng3r committed:
Rewrite ar and dr commands with newshell

Breaking changes:
ar* => arf, ar- => arf-, ar. => ar*
ar, => art, ar? => arq
ar0 => ara0, arb => arab, ars => araS
arC => arpC
arn => arR
art => arT
arf => arF
Removed: drm, aer

ret2libc commented on librz/cons/line.c:
```diff
@@ -162,4 +162,16 @@ RZ_API void rz_line_ns_completion_result_add(RzLineNSCompletionResult *res, cons
+ * \param option the option to be added
+ * \param cur currently entered prefix
+ */
+RZ_API void rz_line_ns_completion_result_propose(RzLineNSCompletionResult *res, const char *option, const char *cur, size_t cur_len) {
```
:+1:

ret2libc commented on librz/core/cmd/cmd_regs.c:
```diff
@@ -0,0 +1,71 @@
+static RzList *filter_reg_items(RzReg *reg, RZ_NULLABLE const char *filter) {
+	rz_return_val_if_fail(reg, NULL);
+	// default
+	if (!filter || !*filter) {
```
```suggestion
	if (RZ_STR_ISEMPTY(filter)) {
```

ret2libc reviewed (commented):
Things look ok so far. I guess you still need to finish the implementation of some handlers where there are TODOs.

thestr4ng3r commented:
Yes, it is currently only a fraction of the commands. I will probably end up rewriting most of them from scratch without looking much at the old code, since the functionality to provide is quite simple and the old code is unnecessarily complex for that.

thestr4ng3r committed:
Show group-level details of commands

Details can be defined at either a group command or its subcommand of
the same name. If define on group-level it will also be shown in the
group help without needing ??.

ret2libc commented on librz/core/cmd/cmd_api.c:
```diff
@@ -1193,6 +1171,30 @@ static void fill_details(RzCmd *cmd, RzCmdDesc *cd, RzStrBuf *sb, bool use_color
 	}
 }
 
+static char *group_get_help(RzCmd *cmd, RzCmdDesc *cd, bool use_color) {
```
Why the move below? If it is not necessary, could you move it back to the original position so that it's clearer what are the differences here?

thestr4ng3r replied:
Just to avoid the forward decl. But sure, I can move it back.

ret2libc commented on librz/core/cmd/cmd_regs.c:
```diff
@@ -0,0 +1,603 @@
+
+	switch (state->mode) {
+	case RZ_OUTPUT_MODE_STANDARD: {
+		RzListIter *iter;
+		RzRegItem *item;
+		char buf[256] = { 0 };
+		rz_list_foreach (ritems, iter, item) {
```
What I tend to do in my handlers is to have the `rz_list_foreach` outside of the switch because it is something that needs to be done in all modes, so it feels just easier to have common code.

ret2libc reviewed (commented):
Could you test if `wv??` `wv1??` and `wv4??` still work?

thestr4ng3r commented:
![Bildschirmfoto 2021-11-29 um 13 27 55](https://user-images.githubusercontent.com/1460997/143868138-be63a683-2842-4502-8652-89ceeaac1e32.png)

ret2libc commented on librz/core/cmd/cmd_regs.c:
```diff
@@ -4,6 +4,33 @@
+	} while (0)
+#define SYNC_WRITE(type) \
+	do { \
+		if (sync_cb) { \
+			if (!sync_cb(core, type, true)) { \
```
```suggestion
		if (sync_cb && !sync_cb(core, type, true)) { \
```

ret2libc commented on librz/core/cmd/cmd_debug.c:
```diff
@@ -5182,3 +5088,14 @@ RZ_IPI int rz_cmd_debug_continue_until(void *data, const char *input) {
 	return RZ_CMD_STATUS_OK;
 }
+
+#define CMD_REGS_PREFIX   debug
+#define CMD_REGS_REG_PATH dbg->reg
+static bool cmd_regs_sync(RzCore *core, RzRegisterType type, bool write) {
+	return rz_debug_reg_sync(core->dbg, type, write);
+}
+#define CMD_REGS_SYNC cmd_regs_sync
+#include "cmd_regs_meta.inc"
+#undef CMD_REGS_PREFIX
+#undef CMD_REGS_REG_PATH
+#undef CMD_REGS_SYNC
```
you haven't added `dr` commands in the yaml files yet, did you? You are going to add them, if I get this right?

thestr4ng3r replied:
Exactly, they are still missing in the yaml. When they are there, their handlers will point to the macro-generated functions here.

XVilka commented on librz/reg/reg.c:
```diff
@@ -161,6 +161,18 @@ RZ_API const char *rz_reg_get_role(int role) {
 }
 
+/// Get the RzRegisterId with the given name or -1
+RZ_API int rz_reg_role_by_name(const char *str) {
```
Add `RZ_NONNULL` to the `str` since it's the API function.

XVilka commented on librz/core/cautocmpl.c:
```diff
@@ -432,6 +432,56 @@ static void autocmplt_cmd_arg_global_var(RzCore *core, RzLineNSCompletionResult
 	rz_list_free(list);
 }
 
+static void autocmplt_cmd_arg_reg_filter(RzCore *core, const RzCmdDesc *cd, RzLineNSCompletionResult *res, const char *s, size_t len) {
```
Check for `s` being NULL maybe?

thestr4ng3r replied:
It has only one caller and if that gets passed a null pointer, it probably wouldn't propagate here.

XVilka commented on librz/core/cautocmpl.c:
```diff
@@ -432,6 +432,56 @@ static void autocmplt_cmd_arg_global_var(RzCore *core, RzLineNSCompletionResult
+		rz_line_ns_completion_result_propose(res, name, s, len);
+	}
+
+	if (reg) {
```
Maybe reverse the logic? `if (!reg) { return }`

XVilka commented on librz/core/cdebug.c:
```diff
@@ -174,37 +174,50 @@ RZ_API bool rz_core_debug_continue_until(RzCore *core, ut64 addr, ut64 to) {
+		return NULL;
+	}
+	int size = rz_analysis_get_address_bits(core->analysis);
+	RzList *ret = rz_list_new();
```
Check if `ret` is `NULL`

XVilka commented on librz/core/cdebug.c:
```diff
@@ -174,37 +174,50 @@ RZ_API bool rz_core_debug_continue_until(RzCore *core, ut64 addr, ut64 to) {
 }
-static void regs_to_flags(RzCore *core, int size) {
+
+RZ_IPI RzList /*<RzRegItem>*/ *rz_core_regs2flags_candidates(RzCore *core, RzReg *reg) {
```
Maybe a brief description of the function?

XVilka commented on librz/core/cmd/cmd_regs.c:
```diff
@@ -0,0 +1,787 @@
+ * \param arg the full argument string, like "reg = 0x42"
+ * \param eq_pos index of the '=' in arg
+ */
+static RzCmdStatus assign_reg(RzCore *core, RzReg *reg, RzCmdRegSync sync_cb, const char *arg, size_t eq_pos) {
```
Add `RZ_NONNULL` and assert for `arg` since there is a `strdup()` call.

thestr4ng3r replied:
There is only one caller which checks null before. But `RZ_NONNULL` makes sense ofc.

XVilka commented on librz/core/cmd/cmd_regs.c:
```diff
@@ -0,0 +1,787 @@
+		SYNC_READ(-RZ_REG_TYPE_FPU);
+		return RZ_CMD_STATUS_OK;
+	}
+	char *name = rz_str_trim_dup(argv[1]);
```
Maybe you could just copy the logic for the newshell parsing from the `e` command handler?

thestr4ng3r replied:
I did check `e` but the logic is a bit different there. The arg here has the `_LAST` flag and supports one assignment with spaces.

XVilka commented on librz/core/cmd/cmd_regs.c:
```diff
@@ -0,0 +1,787 @@
+			long double val = 0.0f;
+#if __windows__
+			double dval = 0.0f;
+			sscanf(eq, "%lf", (double *)&dval);
```
Why not use `rz_num_` API here instead?

thestr4ng3r replied:
It's the copied old broken code. I am still not sure what exactly to do with it.

thestr4ng3r replied:
Ok, I checked the `rz_num` api, but it does not support `long double`, which is necessary here in certain cases. The code is still not optimal, but I think rewriting should be done here entirely.

XVilka commented on librz/core/cmd/cmd_regs_meta.inc:
```diff
@@ -0,0 +1,118 @@
+
+/**
+ * @file
```
Missing SPDX, please also use `\file` syntax.

thestr4ng3r committed:
Fix > $ redirection for errored command and leak

wargio commented on librz/core/cmd/cmd_regs_meta.inc:
```diff
@@ -0,0 +1,124 @@
+ *   CMD_REGS_SYNC       RzCmdRegSync function or NULL
+ */
+
+#define HANDLER__(prefix, name) rz_##prefix##_##name##_handler
+#define HANDLER_(prefix, name)  HANDLER__(prefix, name)
+#define HANDLER(name)           HANDLER_(CMD_REGS_PREFIX, name)
```
@ret2libc how about having this accessible everywhere?

thestr4ng3r replied:
It is very specific to this file.

wargio commented on test/db/archos/linux-x64/dbg_step_back:
```diff
@@ -29,25 +32,35 @@ db @ 0x0040052f
 dsb
 dsb
-dr rax,rbx,rcx,r10,rbp,rip
+dr rax
+dr rbx
+dr rcx
+dr r10
+dr rbp
+dr rip
```
I was thinking, why not allowing dr/ar to have multiple args as array of values?
so this would become `dr rax rbx rcx r10 rbp rip`

wargio replied:
if the problem is when you assign, the problem is non-existent because the arg will still be one 
`dr rax=0x1243 rbx=0x1243 rcx=0x1243 r10=0x1243 rbp=0x1243 rip=0x1243`

thestr4ng3r replied:
We would lose the ability to write `dr rax = 0x1234`, but at the same time it would also be more in line with `e`. Maybe I will refactor out the code from there and reuse it in both, which would also address @XVilka's suggestion from above.

XVilka replied:
Sounds good to me, @ret2libc thoughts?

XVilka replied:
In fact the pattern is general enough, that the parsing of such cases can be generalized and shared between those, and maybe somewhere else in the future.

wargio replied:
ok for me too

XVilka commented:
One more test to update:
```
[XX] C:\projects\rizin\test\db\archos\windows-x64\dbg_dts continue backwards
ANSICON=1 RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e dbg.create_new_console=false
ood
db @ main
3dc
dr rflags=0x200 > nul
dr rflags
dr rax
dts+
7ds
dr rflags
dr rax
dcb
dr rflags
dr rax
doc
' bins/pe/ioli/w32/crackme0x00.exe
-- stdout
--- expected
+++ actual
@@ -1,6 +1,6 @@
-rflags = 0x00000200
-rax = 0x00000001
-rflags = 0x00000216
-rax = 0x0000001e
-rflags = 0x00000200
-rax = 0x00000001
+0x00000200
+0x00000001
+0x00000216
+0x0000001e
+0x00000200
+0x00000001
-- stderr
Spawned new process with pid 5884, tid = 5892
= attach 5884 5892
Try: 'R!pid'
File dbg://C:\\projects\\rizin\\test\\bins/pe/ioli/w32/crackme0x00.exe  reopened in read-write mode
hit breakpoint at: 0x401310
Reading 65536 byte(s) from 0x00010000...
Reading 32768 byte(s) from 0x00098000...
Reading 8192 byte(s) from 0x000b0000...
Reading 61440 byte(s) from 0x000c0000...
Reading 32768 byte(s) from 0x001d8000...
Reading 32768 byte(s) from 0x00384000...
Reading 4096 byte(s) from 0x00405000...
Reading 4096 byte(s) from 0x00406000...
Reading 12288 byte(s) from 0x0060d000...
Reading 45056 byte(s) from 0x00660000...
Reading 151552 byte(s) from 0x00750000...
Reading 4096 byte(s) from 0x00a4f000...
Reading 8192 byte(s) from 0x7527a000...
Reading 12288 byte(s) from 0x763c5000...
Reading 4096 byte(s) from 0x77760000...
Reading 16384 byte(s) from 0x77c45000...
Reading 4096 byte(s) from 0x77c54000...
Reading 24576 byte(s) from 0x77d7e000...
Reading 4096 byte(s) from 0x7ff58000...
Reading 4096 byte(s) from 0x7ff61000...
Reading 4096 byte(s) from 0x7ff80000...
Reading 4096 byte(s) from 0x7ff91000...
Reading 4096 byte(s) from 0x7ffcbba3a000...
Reading 4096 byte(s) from 0x7ffcbdc0a000...
Reading 4096 byte(s) from 0x7ffcbe76f000...
Reading 32768 byte(s) from 0x7ffcbe772000...
hit breakpoint at: 0x401310 cnum: 0
Error while parsing command: `Usage: d   # Debug commands
| db[?]                    Breakpoints commands
| dbt[?]                   Display backtrace based on dbg.btdepth and dbg.btalgo
| dc[?]                    Continue execution
| dd[?]                    File descriptors (!fd in r1)
| de[-sc] [perm] [rm] [e]  Debug with ESIL (see de?)
| dg <file>                Generate a core-file (WIP)
| dH [handler]             Transplant process to a new handler
| di[?]                    Show debugger backend information (See dh)
| dk[?]                    List, send, get, set, signal handlers of child
| dL[?]                    List or set debugger handler
| dm[?]                    Show memory maps
| do[?]                    Open process (reload, alias for 'oo')
| doo[args]                Reopen in debug mode with args (alias for 'ood')
| doof[file]               Reopen in debug mode from file (alias for 'oodf')
| doc                      Close debug session
| dp[?]                    List, attach to process or thread id
| dr[?]                    Cpu registers
| ds[?]                    Step, over, source line
| dt[?]                    Display instruction traces
| dw <pid>                 Block prompt until pid dies
| dW                       List process windows
| dWi                      Identify window under cursor
| dx[?]                    Inject and run code on target process (See gs)
```

XVilka reviewed (approved):
Further improvements can be done in a separate PR.
